hashing/md5.cpp

```diff
@@ -0,0 +1,370 @@
+/**
+ * @file
+ * @author [tGautot](https://github.com/tGautot)
+ * @brief Simple C++ implementation of the [MD5 Hashing
+ * Algorithm](https://en.wikipedia.org/wiki/MD5)
+ *
+ * @details
+ * The [MD5 Algorithm](https://en.wikipedia.org/wiki/MD5) is a
+ * hashing algorithm which was designed in 1991 by [Ronal
+ * Rivest](https://en.wikipedia.org/wiki/Ron_Rivest).
+ *
+ * MD5 is one of the most used hashing algorithm there is. Some of its
+ * use cases are:
+ *  1. Providing checksum for downloaded software
+ *  2. Store salted password
+ *
+ * However MD5 has be know to be cryptographically weak for quite some
+ * time, yet it is still widely used. This weakness was exploited by the
+ * [Flame Malware](https://en.wikipedia.org/wiki/Flame_(malware)) in 2012
+ *
+ * ### Algorithm
+ * First of all, all values are expected to be in [little endian]
+ * (https://en.wikipedia.org/wiki/Endianness). This is especially important
+ * when using part of the bytestring as an integer.
+ *
+ * The first step of the algorithm is to pad the message for its length to
+ * be a multiple of 64 (bytes). This is done by first adding 0x80 (10000000)
+ * and then only zeroes until the last 8 bytes must be filled, where then the
+ * 64 bit size of the input will be added
+ *
+ * Once this is done, the algo breaks down this padded message
+ * into 64 bytes chunks. Each chunk is used for one *round*, a round
+ * breaks the chunk into 16 blocks of 4 bytes. During these rounds
+ * the algorithm will update its 128 bit state (represented by 4 ints: A,B,C,D)
+ * For more precisions on these operations please see the [Wikipedia
+ * aritcle](https://en.wikipedia.org/wiki/MD5#Algorithm).
+ * The signature given by MD5 is its 128 bit state once all rounds are done.
+ *
+ * @note This is a simple implementation for a byte string but
+ * some implmenetations can work on bytestream, messages of unknown length.
+ */
+
+#include <algorithm>  // Used for std::copy
+#include <array>      // To avoid c-style array (thanks clang-tidy)
+#include <cassert>
+#include <cstring>  // Used for std::memcopy
+#include <iostream>
+#include <string>
+#include <vector>  // To avoid c-style array of runtime size (thanks clang-tidy)
```

Comment:
Please add a one-line description of what the library/header is for (**see the example below**).

```c++
#include <iostream>    /// for IO operations
#include <cassert>    /// for assert
```